### PR TITLE
921 env var catalog

### LINKED ISSUE #921: [BUG] Environment variables in capability details page not displayed properly
**Describe the bug**
Environment variables get deleted when editing and submitting a capability

**To Reproduce**
1. Add a capability with multiple environment variables (lets say 3). Submit it.
2. Edit the same capability. (Click edit button). No need to modify any fields. Submit it.
3. View the capability details page. You will see less number of env variables (<3. Probably 2).
4. If you edit and submit again, the number of env variables decreases by 1 each time.

During the edit capability - the env indexes (_i) probably starts again (from 0) and the indexing is different in edit and create page. This might be the issue.

**Expected behavior**
All environment variables need to be present in view capability page (capability-details page). 

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Mac
 - Browser : Chrome
 - Version : Catalog dev branch

**Additional context**

### PR DESCRIPTION
**Pull request by minump**
## Description
Environment variable key and value are indexed by the following pattern : 
- environmentVariables_key_{{num_env}}_{{num_cap}} and 
- environmentVariables_value_{{num_env}}_{{num_cap}} 
where num_env and num_cap are 0 indexed.

**Fixes #921**

[comment]: # (This project only accepts pull requests related to open issues.)
[comment]: # (If suggesting a new feature or change, please discuss it in an issue first.)
[comment]: # (If fixing a bug, there should be an issue describing it with steps to reproduce.)

## Review Time Estimate
_Please give your idea of how soon this pull request needs to be reviewed by selecting one of the options below. This can be based on the criticality of the issue at hand and/or other relevant factors._

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Immediately
- [x] Within a week
- [ ] When possible

## Type of changes
_Please select a relevant option:_

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [x] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Other (any another change that does not fall in one of the above categories.)

## Checklist:
_Please select all applicable options:_

[comment]: # (To select your options, please put an 'x' in the all boxes that apply.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] I have signed the Rokwire Contributor License Agreement ([CLA](https://rokwire.org/rokwire_cla)). (_Any contributor who is not an employee of the University of Illinois whose official duties include contributing to the Rokwire software, or who is not paid by the Rokwire project, needs to sign the CLA before their contribution can be accepted._)
- [x] I have updated the [CHANGELOG](../CHANGELOG.md).
- [ ] I have read the [Contributor Guidelines](../CONTRIBUTING.md).
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] My change requires updating the documentation.
- [ ] I have made necessary changes to the documentation.
- [ ] I have added tests related to my changes.
- [ ] My changes generate no new warnings.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

---

[comment]: # (Template credit: This pull request template is based on Embedded Artistry {https://github.com/embeddedartistry/templates/blob/master/.github/PULL_REQUEST_TEMPLATE.md}, Clowder {https://github.com/clowder-framework/clowder/blob/develop/.github/PULL_REQUEST_TEMPLATE.md}, and TalAter {https://github.com/TalAter/open-source-templates} templates.)

### CODE REVIEW

**Comment by minump** [comment]
updated indexing with children method and filtering. Ready for review.

File: contributions/catalog/models/capability_utilities.py [hunk 1]
```diff
@@ -63,14 +64,15 @@ def to_capability(d):
...
         capability['id'] = cap_id
-
-        env_k, env_v = d['environmentVariables_key_' + str(i)], d['environmentVariables_value_' + str(i)]
-        for k, v in list(zip(env_k, env_v)):
-            capability["deploymentDetails"]['environmentVariables'].append({'key': k, 'value': v})
+        key_pattern = re.compile('environmentVariables_key_[0-9]+' + '_' + str(i))
+        val_pattern = re.compile('environmentVariables_value_[0-9]+' + '_' + str(i))
+        d_keys = list(filter(key_pattern.match, d))  # filter keys matching pattern
+        d_vals = list(filter(val_pattern.match, d))  # filter keys matching pattern
+        for k, v in zip(d_keys, d_vals):
```
Comment (sandeep-ps) [inline contributions/catalog/models/capability_utilities.py:74]: Please add some comments explaining the logic in this method. Thanks.

File: contributions/catalog/webapps/templates/contribute/contribute.html [hunk 2]
```diff
@@ -581,10 +581,10 @@ <h3 id="capability">Add a Capability to this Contribution</h3>
                     <div name="add-env-var-div" class="form-group">
                         <label class="required control-label col-sm-2">Environment Variable(s)</label>
                         <div class="col-sm-10" id="key-val">
-                            <input class="form-control" name="environmentVariables_key" placeholder="Key"
-                                   id="environmentVariables_key" type="text">
-                            <input class="form-control" name="environmentVariables_value" placeholder="Value"
-                                   id="environmentVariables_value" type="password">
+                            <input class="form-control" name="environmentVariables_key_0" placeholder="Key"
```
Comment (sandeep-ps) [inline contributions/catalog/webapps/templates/contribute/contribute.html:584]: Similarly, please add some comments about how the environment variable ID indexes are searched and updated. Thanks.

**Comment by sandeep-ps** [comment]
I tested this by adding a Contribution and updating an environment variable and its values as described in the issue. It seems to be working fine. I have left some comments for improving code readability. The rest looks good to me. Thanks.

**Comment by sandeep-ps** [comment]
@ywkim312, Please share if you have any feedback on this PR. Thanks.

**Comment by ywkim312** [comment]
@minump Can you give little more explanation of what is this PR for?

**Comment by minump** [comment]
> @minump Can you give little more explanation of what is this PR for?

the bug is described [here](https://github.com/rokwire/rokwire-building-blocks-api/issues/921). The fix is by correcting the envvar pair indexing with pattern environmentVariables_key_envindex_capindex

File: contributions/catalog/models/capability_utilities.py (hunk 1 above)
Comment (minump) [reply contributions/catalog/models/capability_utilities.py:76]: updated

**Comment by ywkim312** [comment]
@minump @sandeep-ps I was able to reproduce the bug. Sorry about the above comment please ignore it.

**Review by ywkim312 — APPROVED** [review]
It worked so I approve

**Review by sandeep-ps — APPROVED** [review]

**Comment by sandeep-ps** [comment]
Thank you! Approved and merged.

**Pull request merged**